File: java/src/com/sortcompare/QuickSort.java
package com.sortcompare;

import java.util.*;

public class QuickSort {

    enum PivotType {RANDOM, FIRST, LAST}

    static ArrayList<Integer> sort(final ArrayList<Integer> unsortedList, final PivotType pivotType) {

        if (unsortedList.size() < 2) {
            return unsortedList;

        } else {
            int pivot = unsortedList.remove(0);
            final ArrayList<Integer> left = new ArrayList<>();
            final ArrayList<Integer> right = new ArrayList<>();

            for (Integer item : unsortedList) {
                if (item <= pivot) {
                    left.add(item);
                } else {
                    right.add(item);
                }
            }
            final ArrayList<Integer> sortedList = new ArrayList<>();
            sortedList.addAll(sort(left, pivotType));
            sortedList.add(pivot);
            sortedList.addAll(sort(right, pivotType));
            return sortedList;
        }
    }
}

File: kotlin/src/sortcompare/QuickSort.kt
package sortcompare

import java.util.*

class QuickSort {

    companion object {
        enum class PivotType { FIRST, LAST, RANDOM}

        fun IntRange.random() =
                Random().nextInt((endInclusive + 1) - start) +  start

        fun sort(unsortedList: ArrayList<Int>, pivotType: PivotType): ArrayList<Int> {
            if(unsortedList.size < 2) {
                return unsortedList
            }
            else {
                val i: Int = when(pivotType) {
                    PivotType.FIRST -> 0
                    PivotType.LAST -> unsortedList.size - 1
                    PivotType.RANDOM -> (0 until unsortedList.size).random()
                }
                val pivot = unsortedList.removeAt(i)

                val left = unsortedList.filter { it <= pivot } as ArrayList<Int>
                val right = unsortedList.filter { it > pivot } as ArrayList<Int>
                val sortedList: ArrayList<Int> = ArrayList()
                sortedList.addAll(sort(left, pivotType))
                sortedList.add(pivot)
                sortedList.addAll(sort(right, pivotType))
                return sortedList
            }
        }
    }
}